### ISSUE TITLE
Issues with the Typelevel Cats library

### ISSUE THREAD

**Issue by bfil** [ensime/ensime-server#1525]
### Steps to reproduce

1. `git clone https://github.com/bfil/ensime-cats-bug`
2. open the project and your editor of choice (I tried with Atom and Sublime) and start Ensime

### Expected behaviour

No issues should be shown in the editor, the project compiles just fine, it also works just fine in the Eclipse Scala IDE.

### Actual behaviour

The editors show a few compiler errors, here's a screenshot taken from my Atom editor

![screen shot 2016-06-20 at 16 11 10](https://cloud.githubusercontent.com/assets/660513/16199613/2eb7fcf6-3702-11e6-9a75-5f2ab96c64ba.png)

**Comment by fommil** [comment]
this is the FAQ #673 https://github.com/fommil/imaginary-friend

**Comment by bfil** [comment]
Sorry @fommil, how can I test if that is actually coming from the scala presentation compiler?

**Comment by bfil** [comment]
Also, Scala IDE (Eclipse) uses the presentation compiler afaik, but the project there doesn't show any error.

**Comment by fommil** [comment]
and you compiled the project?

**Comment by bfil** [comment]
I sure did

**Comment by fommil** [comment]
and it still didn't work after you compiled the project?

**Comment by bfil** [comment]
Yep, I usually generate the `.ensime` file after compiling the project.

Tried again, it looks like it works at first, but as soon as I add a blankspace somewhere then it shows those errors.

If I stop ensime, restart it, it shows no errors, even if I trigger manually the typecheck from the editor a few times, all fine, until I change the code (even a blankspace somewhere), that makes everything go bust, and it does never show it without errors no matter how many times I trigger the typecheck.

But making sure it's compiled first fixes the issue in Sublime, so this might be related to ensime-atom.

Would you suggest opening the issue there instead?

**Comment by fommil** [comment]
yeah it sounds like an atom issue (that's usually the best place to create tickets). I'll try to reproduce in Emacs. Ping me in a few days if I don't respond.

**Comment by bfil** [comment]
Ok, thanks. Raised it here because I saw it happening in Sublime as well, but compiling first fixes it there, I'll open the issue there then, but keep me posted, trying to reproduce it should just be a `git clone` away.

**Comment by hedefalk** [comment]
@fommil I'm pretty sure it's triggered by Atom asking for Implicits (+ typechecking the file simultanously). If I just disable the call for ImplicitInfo the problem goes away. This has been around for ages, but I haven't come around to digging further for some reason but will now. I'll be back!

**Comment by fommil** [comment]
Oh damn

**Comment by bfil** [comment]
Has anyone had a chance to look at this issue?

**Comment by fommil** [comment]
@bfil if you wait until after scala world, we're making a library that you can use to reproduce and investigate further. I wouldn't hold your breath for somebody to fix this for you though, you'll have to get your hands dirty.

**Comment by bfil** [comment]
I wanted to give it a try myself, some basic advice would be helpful though, it looks like a server issue to me since I get the same errors on both Ensime and Sublime. But would be great knowing where should I start looking to fix this.

**Comment by fommil** [comment]
wait until after scala world and I'll have some tools for you to use.

**Comment by bfil** [comment]
@fommil sounds great, will wait for the library. Shouldn't this issue be re-opened?

**Comment by fommil** [comment]
no, issues with the presentation compiler will be tracked elsewhere it's not an ensime bug. This is a dupe of #673

**Comment by bfil** [comment]
@fommil I thought it wasn't a presentation compiler bug since Eclipse is fine with my cats project, and Eclipse is using the presentation compiler afaik

**Comment by fommil** [comment]
until we have a way of testing this kind of stuff I'm assuming it's all presentation compiler related. You could write an ensime specific test if you wanted by following the contributor guide and starting with something like RichPresentationCompilerSpec and pick a project base with cats in it. I think we already have a test project with cats in it.

**Comment by bfil** [comment]
Will have a look

**Comment by hedefalk** [comment]
I'm interested to follow this too… It's my impression that a lot of the presentation compiler bugs we have seems to be worked around in scala-ide. I know for a fact that I have had a lot of false positives on projects in Ensime that wasn't there in Eclipse. Exactly how this is done, I've been meaning to ask. It just feels that problems in presentation compiler hits harder in Ensime than in scala-ide. Maybe @dragos can tell us about workarounds or if I have been imagining this?

**Comment by dragos** [comment]
There's no specific work-around for such errors, but there's a strict thread discipline to avoid any concurrent queries to the PC. I will be using ensime a lot more once I can get it to work in VS Code (I have the language server, now I need to plug-in a backend, either ensime or the PC directly), so I'll keep an eye on such errors. I am not familiar with the ensime code-base yet, but I think there are some peculiar ways to treat in-memory files so that's where I'd look first.

**Comment by fommil** [comment]
FYI the tooling I'm referring to is https://github.com/ensime/pcplod

**Comment by hedefalk** [comment]
@dragos Have you that language server implemention up anywhere I can look? It's always good to look at the options…

For what it's worth I have been toying a bit more with vscode using the programmatic language api and ensime via my ensime-client lib (same as atom). I have some crude features working:

* definitions (including navigation)
* type hover
* completions
* errors/warnings

Still lots to cleanup (code is ugly), but it feels promising, think it will be on par with Atom soon. Just need those rows/cols: https://github.com/ensime/ensime-server/pull/1598

Sorry to hijack thread…

**Comment by dragos** [comment]
Sure: https://github.com/dragos/dragos-vscode-scala

The language server does not implement the whole protocol yet (although all types are defined). It's probably very close. I might refactor some things, but the rough idea is there. The client spawns and communicates to the server (when saving a file an "alert" is sent back to VS Code).

**Comment by hedefalk** [comment]
I'm pretty sure that issues like this: https://github.com/ensime/ensime-atom/issues/235 arise from the fact that ensime-atom posts more questions to ensime-server out-of-the-box and that they haven't been noticed that much in ensime-emacs because the client behaves "better".

But to me it sounds like an underlying problem that we should probably adress the way scala-ide does it?

* @dragos So you basically lock on the PC and only asks one thing at a time? Is it a "known fact" that the presentation compiler doesn't work with multiple concurrent requests?

* @fommil @rorygraves I guess ensime doesn't do this, but just lets whatever the client asks for be passed through then? 

Should we change this?

So just to make clear how I see it: 
I have an open ensime-atom bug here https://github.com/ensime/ensime-atom/issues/235 which is just a dupe of this one. This one here is closed since this is marked as upstream bug in PC. I've kept the ensime-atom issue open since it's not a problem in ensime-emacs, but on the other hand, we basically know that it's all triggered by multiple queries that atom posts so:

Either:

1) It's a bug in the presentation compiler which SHOULD handle concurrent requests.

2) It's a known fact that the presentation compiler DOESN'T handle concurrent requests and it's up to users to have a locking policy. In this case I suggest it's up to ensime-server to handle it and this issue should be re-opened.

3) It's not ensime-server's fault either, but up to the editor client to not post multiple concurrent requests - then it's a bug in ensime-atom and I'll try to fix it. But it feels this will be very hard to do in a performant way. Some requests to ensime-server aren't even presentation compiler related (docs for instance) so then I have to start sorting these out client side which feels weird?

4) We should really support multiple requests but there are just some corner cases and the pessimistic view of scala-ide is not good enough - then we probably need to identify which cases things blow up and just avoid concurrency on just these cases…?

**Comment by fommil** [comment]
The ensime server also makes sure that requests to the PC are single threaded. That's why it is managed as an actor. I really doubt that this is the cause of the problem.

I don't want to jump to conclusions about what is going wrong without being able to reliably reproduce these bugs and that's why we are building `pcplod`. You can also create ensime tests and if you can reproduce there but NOT in `pcplod` then we should reopen this ticket.

We might try to put all these sorts of bugs, and attempts to fix them, in the `ensime/scala` repository. We'll have to think about that. We might be able to use the `ensime-monkeys` mechanism to roll out fixes to everybody, depending on the nature of the fix.

That said, we still need contributors who are prepared to get their hands dirty and reproduce/fix these bugs. Once the infrastructure is in place, maybe Sunday-ish, then we can ask `cats` people to help out.

**Comment by hedefalk** [comment]
Ok, cool! I'll try to see if I can make a standalone ensime test for it, I'll start there…

**Comment by hedefalk** [comment]
@fommil Sorry to ask, but if I would like to try create an ensime test that exposes this issue, do you think you can point me to a physical location where I should write such a test?

**Comment by bfil** [comment]
@hedefalk I've forked the server and started hacking a test together, I have no idea what I'm doing either, for now I've created a `CatsSpec` under `core/src/it/scala/org/ensime/core`.

I started with a simple test that uses the `askTypeInfoAt` to get the type of a value, and it looks like it's unable to resolve the `cats.data.XorT` type for some reason if I create a simple value using `XorT.right`, but it always gives me back a `BasicTypeInfo("<none>", Nil, "<none>")`.

The issue is that this kind of test does not give me any info over what's going wrong, so yeah, no idea what I'm doing, but I've noticed that if I define my own `XorT` type with the same signature of cats `askTypeInfoAt` returns the expected type. So the must be something going wrong when ensime is parsing `cats.data.XorT` imo. In the project I shared, ensime server throws a bunch of weird logs in relation to `XorT` around the `?` in the type parameters, but I'm not entirely sure if they're relevant..

If you want to take a look at the source code: `XorT` has now been renenamed to `EitherT`..
https://github.com/typelevel/cats/commit/3bc775fcfe53609c384c877665be179560c8b1be

**Comment by bfil** [comment]
Here's what I've done so far:
https://github.com/bfil/ensime-server/commit/f5afd7c95ae9a33d2adbd56fdb35afa95625a9f8

The `TypelevelLibrariesSymbolToFqnSpec` integration test is failing for some reason after my changes, probably due to the cats dependency version?

My integration test to check the type info seems to be passing, not sure where to go from here, or how to trigger the `NoteError`s that I see in the Ensime server logs when going to the definition of `XorT`/`EitherT`.

**Comment by fommil** [comment]
Sounds like you're on the right tracks. We should have a test template project somewhere with cats on it. Might be in the graph branch.

**Comment by hedefalk** [comment]
Cool!

**Comment by fommil** [comment]
it's this one https://github.com/ensime/ensime-server/blob/2.0/core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala#L84-L87

**Comment by bfil** [comment]
Thanks @fommil 

Yeah, that's the one I'm using in my `CatsSpec`.

**Comment by bfil** [comment]
Any clue why if I change the `cats` dependency in `EnsimeBuild.scala` to `intransitive()` then `TypelevelLibrariesSymbolToFqnSpec` works but my `CatsSpec` fails (cannot resolve cats), while if I remove `intransitive()` my `CatsSpec` passes but then `TypelevelLibrariesSymbolToFqnSpec` fails because it tries to index some `scalacheck` code but fails in doing so?

@hedefalk have you had any luck into trying to reproduce the issues on this side? Cause apart from the dependency woes it looks like indexing and typechecking works just fine within ensime server.. Of course I am not sure how to try out your theory about the concurrent calls, and which calls those would be.

**Comment by fommil** [comment]
@bfil @hedefalk we now have pcplod released! Could one of you please submit a PR to cats to add pcplod as a test dependency and write a couple of pieces of idiomatic cats code that causes problems, with tests. It's super easy to use, much easier than writing an ensime test https://github.com/ensime/pcplod

Also, one thing I noticed that might cause problems is if you have compiler plugins in some subprojects but not the root project, which means ensime doesn't get them... and then you'll get a lot of problems with macros especially if you're using macro-paradise. It's an epic to fix that one between https://github.com/ensime/ensime-sbt/issues/217 and https://github.com/ensime/ensime-server/issues/1152 and is the next priority for @sugakandrey after the graphpocalypse is merged into 2.0

**Comment by bfil** [comment]
@fommil I've forked cats and I'm trying to create a basic initial test but it seems like `pcplod` cannot find my Scala files in the resources folder, so I'm stuck at square one.

**Comment by fommil** [comment]
@rorygraves can you please help out @bfil ? (I'm AFK for a while)

**Comment by bfil** [comment]
No prob @fommil I can push what I have in my `cats` fork if that helps.

**Comment by rorygraves** [comment]
I will look tomorrow night. Ping me on gitter if I forget.

**Comment by hedefalk** [comment]
@bfil Sorry, I haven't had any time to spare. My idea was to just to look at the request/response logs from ensime-atom and do just the same "headless". My client test suggested that a concurrent ImplicitInfoReq just before typechecking somehow triggers the problem.

@fommil Awesome work!

**Comment by bfil** [comment]
@hedefalk oh ok, an headless integration test might be easier.

I've finally managed to make `pcplod` work in the `cats` code base, I move the settings from the tests module to the jvm module and there it finds the test resources just fine, no clue why.. (might be some obscure reason due to the `cats` cross compilation to both JVM and JS).

I'll keep working on it, will give an update if I manage to make the presentation compiler blow.

**Comment by bfil** [comment]
So, looking at the `?` in type parameters in `EitherT`, which seems to be causing `ensime-atom` and `ensime-sublime` to blow up once the `EitherT.scala` file is analyzed, it looks like `cats` use the `kind-projector` compiler plugin to facilitate writing type lambdas, but it is in its common settings which are applied to each SBT module/project, it also uses macro paradise.

These are a subset of the logs in regards to the `?` in `EitherT.scala` from ensime.log in Atom (scroll to the far right):

```
22:39:23.410 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,not found: type ?,NoteError,2915,2916,71,80)))
22:39:23.410 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,Either[A,<error>] takes no type parameters, expected: one,NoteError,2905,291
22:39:23.515 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,not found: type ?,NoteError,4371,4372,105,74)))
22:39:23.515 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,Either[A,<error>] takes no type parameters, expected: one,NoteError,4361,437
22:39:23.517 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,missing argument list for method apply in object EitherT Unapplied methods a
22:39:23.595 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,not found: type ?,NoteError,9179,9180,209,37)))
22:39:23.596 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,Either[A,<error>] takes no type parameters, expected: one,NoteError,9169,918
22:39:23.599 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,Either[A,<error>] takes no type parameters, expected: one,NoteError,9169,918
22:39:23.618 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,not found: type ?,NoteError,9208,9209,209,66)))
22:39:23.619 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,Either[A,<error>] takes no type parameters, expected: one,NoteError,9198,921
22:39:23.743 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,not found: type ?,NoteError,12347,12348,285,70)))
22:39:23.751 DEBUG akka://ENSIME/user/$a o.e.s.ConnectionHandler - received handled message NewScalaNotesEvent(false,List(Note(/ensime-cats-bug/.ensime_cache/dep-src/source-jars/cats/data/EitherT.scala,<error> forSome { type _$1 } takes no type parameters, expected: one,NoteErr
```

Using `pcplod` I didn't find a way to replicate this behaviour (in an editor you just need to go to the `EitherT` definition and Ensime reports a bunch of `NewScalaNotesEvent` about the `?` occurences), I'm not entirely sure how I would add a test using `pcplod` for it in cats to highlight this, cause the rest (fetching symbols and types) seems to be working fine.

**Comment by bfil** [comment]
Also found this 1 year old Gitter discussion about `kind-projector`: https://gitter.im/ensime/ensime-server/archives/2015/09/30

Since the presentation compiler seems happy with the test I've added using `pcplod` in `cats`, I'm assuming this issue is an Ensime one, I would like to try and add a test in `ensime-server` that uses `kind-projector` as a compiler plugin, but it might be too advance for me to setup without any guidance.

The only compiler plugin in this codebase I can see is part of the shapeless dependency for Scala 2.10. I'll look at that to see if I can use it as a starting point to write a test.

Sorry to keep spamming, but hopefully this can help.

**Comment by bfil** [comment]
So, just found out that if I add this to my project build (the one that has a dependency on `cats`):

```
addCompilerPlugin("org.spire-math" %% "kind-projector" % "0.9.0")
```

The errors related to `cats` in Atom just disappear, so it is definitely a compiler plugin related issue, like @fommil suggested.

This makes me quite happy, it means that I can just add the compiler plugin to my projects depending on `cats` as a workaround, and enjoy a working editor!

**Comment by fommil** [comment]
This is not spam!!! Great investigations. Make sure to add any plugins your project uses to the root project to make sure ensime sees it, were working on fixing this.

I'm on holiday for 3 weeks so sorry I can't help.

**Comment by hedefalk** [comment]
Great catch!

**Comment by bfil** [comment]
@fommil no worries, at least we have a workaround.

I do not fully understand what you mean with "add any plugin to the root project", if the compiler plugin is used by a dependency (like in the case of `cats`), the only thing I can do on my side is adding it to my project. Not sure how Ensime handles dependencies' compiler plugins, is the issue that:

1. `cats` is not defining it on its root project
2. I have to define it on my side because the compiler flag for the plugin is not added to the `.ensime` file when a dependency uses a compiler plugin?

**Comment by DStranger** [comment]
Has this issue been fixed?
Looks like it still exists for 2.0.0-M3
